engines/sci/graphics/controls32.cpp

```diff
@@ -353,4 +366,303 @@ void GfxControls32::flashCursor(TextEditor &editor) {
+		s += Common::String::format("|f%d|", line._fontId);
+	if (line._foreColor != -1)
+		s += Common::String::format("|c%d|", line._foreColor);
+	if (line._alignment != -1)
+		s += Common::String::format("|a%d|", line._alignment);
+	s += str;
+
+	line._str = s;
+	_text += s;
+
+
+	if (scrollTo)
+		_firstVisibleChar = _text.size() - s.size();
+
+	update(true);
+
+	return line._id;
+}
+
+void ScrollWindow::upArrow() {
+	if (_topVisibleLine == 0)
+		return;
+
+	_topVisibleLine--;
+	_bottomVisibleLine--;
+
+	if (_bottomVisibleLine - _topVisibleLine + 1 < _numVisibleLines)
+		_bottomVisibleLine = _numLines - 1;
+
+	_firstVisibleChar = _startsOfLines[_topVisibleLine];
+	_lastVisibleChar = _startsOfLines[_bottomVisibleLine+1] - 1;
+
+	_visibleText = Common::String(_text.c_str() + _firstVisibleChar, _text.c_str() + _lastVisibleChar + 1);
+
+	// TODO: Double check this -1 at the end (for the \n)
+	Common::String lineText(_text.c_str() + _startsOfLines[_topVisibleLine], _text.c_str() + _startsOfLines[_topVisibleLine+1] - 1);
+
+	debugC(3, kDebugLevelGraphics, "ScrollWindow::upArrow: top: %d, bottom: %d, num: %d, numvis: %d, lineText: %s", _topVisibleLine, _bottomVisibleLine, _numLines, _numVisibleLines, lineText.c_str());
+
+	_gfxText32->scrollLine(lineText, _numVisibleLines, _foreColor, _alignment, _fontId, kScrollUp);
+
+	if (_visible) {
+		assert(_screenItem);
+
+		_screenItem->update();
+		g_sci->_gfxFrameout->frameOut(true);
+	}
+}
+
+void ScrollWindow::downArrow() {
+	if (_topVisibleLine + 1 >= _numLines)
+		return;
+
+	_topVisibleLine++;
+	_bottomVisibleLine++;
+
+	if (_bottomVisibleLine + 1 >= _numLines)
+		_bottomVisibleLine = _numLines - 1;
+
+	_firstVisibleChar = _startsOfLines[_topVisibleLine];
+	_lastVisibleChar = _startsOfLines[_bottomVisibleLine+1] - 1;
+
+	_visibleText = Common::String(_text.c_str() + _firstVisibleChar, _text.c_str() + _lastVisibleChar + 1);
+
+	Common::String lineText;
+	if (_bottomVisibleLine - _topVisibleLine + 1 == _numVisibleLines) {
+		// TODO: Double check this -1 at the end (for the \n)
+		lineText = Common::String(_text.c_str() + _startsOfLines[_bottomVisibleLine], _text.c_str() + _startsOfLines[_bottomVisibleLine+1] - 1);
+	} else {
+		// scroll in empty string
+	}
+
+	debugC(3, kDebugLevelGraphics, "ScrollWindow::downArrow: top: %d, bottom: %d, num: %d, numvis: %d, lineText: %s", _topVisibleLine, _bottomVisibleLine, _numLines, _numVisibleLines, lineText.c_str());
+
+
+	_gfxText32->scrollLine(lineText, _numVisibleLines, _foreColor, _alignment, _fontId, kScrollDown);
+
+	if (_visible) {
+		assert(_screenItem);
+
+		_screenItem->update();
+		g_sci->_gfxFrameout->frameOut(true);
+	}
+}
+
+
+void ScrollWindow::getLineIndices() {
+	_gfxText32->setFont(_fontId);
+	// set _gfxText32 foreColor, alignment?
+
+	if (_gfxText32->_font->getHeight() != _pointSize) {
+		error("ScrollWindow font size mismatch");
+	}
+
+	Common::Rect r(0, 0, _rect10.width() + 1, _pointSize + 3);
+
+	_startsOfLines.clear();
+
+	int index = 0;
```

Comment:
Use `for` here instead of `while`, since `index` is never used outside of the loop?